### PR TITLE
CSV export for LMS

### PR DESCRIPTION
**Pull request by bradmerlin**
Adds CSV export + download buttons to the Survey and Poll XBlocks.

**JIRA tickets**: [OC-2981](https://tasks.opencraft.com/browse/OC-2981) [MCKIN-5433](https://edx-wiki.atlassian.net/browse/MCKIN-5433)

**Testing instructions**:

1. In a course with a survey or poll xblock,
1. Click on the export button to queue a CSV export
1. Click the now-active download button to download the CSV
1. Notice the difference in formatting between survey and poll output

Screenshot:

screen shot 2017-09-15 at 4 10 24 pm

**Reviewers**
- [ ] @cgopalan  
- [ ] @bradenmacdonald

### CODE REVIEW

File: tests/unit/test_xblock_poll.py [hunk 1]
```diff
@@ -62,7 +65,23 @@ def test_student_view_user_state_handler(self):
...
+        expected_response = {
+            u'choice': None,
+            u'submissions_count': 5,
+            u'tally': {'R': 0, 'B': 0, 'G': 0, 'O': 0},
+        }
+
+        raise
 
+@patch('courseware.models.StudentModule')
+@patch('instructor_task.models.ReportStore')
```
Comment (bradenmacdonald) [inline tests/unit/test_xblock_poll.py]: You shouldn't need to patch the existing tests... I'll explain elsewhere to move the imports so this isn't necessary.

File: poll/poll.py [hunk 2]
```diff
@@ -24,11 +24,14 @@
...
 import functools
 import json
+import time
 
 from markdown import markdown
 import pkg_resources
 from webob import Response
 
+from courseware.models import StudentModule
+from instructor_task.models import ReportStore
```
Comment (bradenmacdonald) [inline poll/poll.py]: Please move these imports inside the methods that use them, so that this XBlock can still work from the `xblock-sdk`/`workbench`.

File: poll/poll.py [hunk 3]
```diff
@@ -96,6 +100,18 @@ class PollBase(XBlock, ResourceMixin, PublishEventMixin):
...
+        # IF we are sill waiting for it to finish
+        default="",
+        scope=Scope.user_state_summary,
+    )
+    last_export_result = Dict(
+        # The info dict returned by the most recent successful export.
+        # If the export failed, it will have an "error" key set.
+        default=None,
+        scope=Scope.user_state_summary,
+    )
```
Comment (bradenmacdonald) [inline poll/poll.py]: Can you please move these fields and all the new methods into a mixin class, to keep the export-related code separate from the rest of the implementation?

File: poll/poll.py [hunk 4]
```diff
@@ -300,6 +316,89 @@ def wrapper(self, request_json, suffix=''):
...
+        if task_result.successful():
+            if isinstance(task_result.result, dict) and not task_result.result.get('error'):
+                self.last_export_result = task_result.result
+            else:
+                self.last_export_result = {'error': u'Unexpected result: {}'.format(repr(task_result.result))}
+        else:
+            self.last_export_result = {'error': unicode(task_result.result)}
+
+    @XBlock.json_handler
+    def get_status(self, data, suffix=''):
```
Comment (bradenmacdonald) [inline poll/poll.py]: Please rename this (and the related private methods and javascript code) to `get_export_status` or something less generic, and add a docstring - I would expect "get_status" to return the result of the poll or something.

File: poll/poll.py [hunk 5]
```diff
@@ -300,6 +316,89 @@ def wrapper(self, request_json, suffix=''):
...
+    def csv_export(self, data, suffix=''):
+        """
+        Asynchronously export given data as a CSV file.
+        """
+        # Launch task
+        from .tasks import export_data as export_data_task  # Import here since this is edX LMS specific
+
+        # Make sure we nail down our state before sending off an asynchronous task.
+        async_result = export_data_task.delay(
+            self.prepare_data(),
```
Comment (bradenmacdonald) [inline poll/poll.py]: By calling `prepare_data()` here, most of the work is being done synchronously, before we even queue the asynchronous task. All that the asynchronous task is doing is writing the results to a CSV file, which is a trivial task that's not worth making asynchronous.

You should instead be calling `export_data_task` with only one argument (the block ID), and within that task is where we should be preparing the data, which is the part that we want to make asynchronous.

File: poll/poll.py (hunk 2 above)
Comment (cgopalan) [reply poll/poll.py]: Also prefix the imports with `lms.djangoapps..` otherwise they give errors

**Comment by cgopalan** [comment]
@bradmerlin Will the "export" and "download" buttons be only visible until the survey and poll are unanswered in the LMS? They disappear as soon as the results are submitted.
Also when I try to download the csv for a survey by clicking the button I get this:
```
http://localhost:8000/media/5a0525aab94db835a763cf62557ec6c79c7ae04f/survey-data-export-2017-09-23-235733.csv
"/edx/var/edxapp/uploads/5a0525aab94db835a763cf62557ec6c79c7ae04f/survey-data-export-2017-09-23-235733.csv" does not exist
```

**Comment by bradenmacdonald** [comment]
@cgopalan That error is normal on devstacks, but see if the file actually exists within the devstack at the given file path.

**Comment by cgopalan** [comment]
@bradenmacdonald @bradmerlin I did not see a file in that path. I realized it was because this uses celery and I had not started celery. When I started celery, it seems that it does not recognize the new task added here `export_csv_data`. I started celery using this:
`paver devstack lms --settings=devstack_with_worker`
and `./manage.py lms celery worker --settings=devstack_with_worker`.
What am I doing wrong?

**Comment by bradmerlin** [comment]
@cgopalan You aren't doing anything wrong – you'll need to add `poll` to your edx-platform's `OPTIONAL_APPS` – I've created a PR for that [here](https://github.com/edx-solutions/edx-platform/pull/897).

**Comment by bradmerlin** [comment]
@cgopalan and with regards to this:

> Will the "export" and "download" buttons be only visible until the survey and poll are unanswered in the LMS? They disappear as soon as the results are submitted.

I chose to go with the decision made for the 'View Results' button. I agree that it's a bit confusing – I assumed there might have been some reason for it.

**Comment by bradmerlin** [comment]
@cgopalan I've improved the UI in [this commit](https://github.com/open-craft/xblock-poll/pull/35/commits/f57f6df5abd11e53f3a5aa3af8621417f754b6f6):

* The export and download buttons don't disappear if you submit a poll / survey
* Nor do they disappear if you click on 'View Results'
* Once the results have successfully been exported, the export button is disabled (until the page is refreshed)

**Comment by cgopalan** [comment]
@bradmerlin some errors from testing:
1. If survey results have not been submitted yet, and I export, the file is shown like this:
```
user_id,username,user_email,Are you enjoying the course?,Do you think you will learn a lot?,Would you recommend this course to your friends?
14,cgopalan,chandrakant@opencraft.com
```

2. If I click export on one of the poll/surveys, the buttons change for both as shown in screenshot below:

screen shot 2017-09-27 at 1 19 37 pm

**Comment by cgopalan** [comment]
@bradmerlin thanks for fixing the above 2 errors. I verified that they work now.
2 more questions:
1. Should the download/export work in studio? The buttons are visible but when I click "Export" I get this error:
`"'Settings' object has no attribute 'GRADES_DOWNLOAD'"`
2. If the user hits "Export" , the export button is disabled and the "download" button is enabled. But if the user has not submitted the answer yet, and submits and then wants to export again, the export button is disabled. Maybe the export button should be enabled when user hits submit?

**Comment by bradmerlin** [comment]
@cgopalan It's not meant to run in Studio – `GRADES_DOWNLOAD` is only available in LMS. I think that people getting results would only be staff-level users, so they would probably not need to export.

**Comment by cgopalan** [comment]
@bradmerlin is there a way to hide them in studio?

**Comment by bradmerlin** [comment]
@cgopalan I think what's visible in studio is just meant to be a visual demo – right?

**Comment by bradenmacdonald** [comment]
@bradmerlin  The Problem Builder Instructor Tool has an example of showing different messaging within Studio vs. the LMS

**Comment by cgopalan** [comment]
@bradmerlin let me know when you have addressed the above 2 points and I will review again

**Comment by cgopalan** [comment]
@bradenmacdonald @bradmerlin just wanted to make sure - is it the right practice to hide the buttons that are non-functional in studio? Or do something else, like show a message?

**Comment by bradmerlin** [comment]
@cgopalan Done and done!

File: tests/unit/test_xblock_poll.py [hunk 6]
```diff
@@ -1,5 +1,6 @@
 import unittest
 import json
+
```
Comment (cgopalan) [inline tests/unit/test_xblock_poll.py]: @bradmerlin please undo this change since it doesnt add anything.

**Comment by cgopalan** [comment]
:+1:

- [x] I tested this: Tested this functionality on LMS. Ensured that a staff user can export CSV files in the LMS for polls and surveys. Not allowed for non-staff users. In Studio, a message is shown on the unit that says CSV files can be exported in the LMS.
- [x] I read through the code
- [x] ~~I checked for accessibility issues~~
- [x] ~~Includes documentation~~

@bradmerlin this looks good. Once you squash remove the extra blank line in the test file and squash the commits, I can merge.

@bradenmacdonald any concerns from your end?

**Comment by bradmerlin** [comment]
@cgopalan Although the extra blank line doesn't add any functionality, it does improve the [file's style](http://edx.readthedocs.io/projects/edx-developer-guide/en/latest/style_guides/python-guidelines.html#imports-order).

File: poll/public/html/poll.html [hunk 7]
```diff
@@ -57,11 +57,21 @@ <h3 class="poll-header">{% trans 'Feedback' %}</h3>
...
-  </div>
 </div>
+{% if can_view_private_results %}
+  {% if not studio_edit %}
+    <div class="export-results-button-wrapper">
+      <button class="export-results-button">Export results to CSV</button>
+      <button disabled class="download-results-button">Download CSV</button>
+    </div>
+  {% else %}
+    <p>Student data and results CSV available for download in LMS.</p>
```
Comment (bradenmacdonald) [inline poll/public/html/poll.html]: Nit: should be "...in **the** LMS"

File: poll/public/html/survey.html [hunk 8]
```diff
@@ -69,7 +69,19 @@ <h3 class="poll-header">{% trans 'Feedback' %}</h3>
...
     </div>
 </div>
+{% if can_view_private_results %}
+  {% if not studio_edit %}
+    <div class="export-results-button-wrapper">
+      <button class="export-results-button">Export results to CSV</button>
+      <button disabled class="download-results-button">Download CSV</button>
+    </div>
+  {% else %}
+    <p>Student data and results CSV available for download in LMS.</p>
```
Comment (bradenmacdonald) [inline poll/public/html/survey.html]: Same Nit: should be "...in **the** LMS"

**Comment by bradenmacdonald** [comment]
@bradmerlin The UI looks nice, and you've done a great job integrating it into Studio and the LMS and showing it only to staff users.

However, Before I had added `poll` to `OPTIONAL_APPS` and spawned a worker: When I click "Export Results to CSV", the block immediately enables the "Download CSV" button, even though the CSV is not ready - and the link it has is of course invalid. I would expect it to show an error message instead, or at least wait forever for the CSV to be generated, with both the "Export Results to CSV" and "Download CSV" buttons disabled until the CSV has been successfully generated, and then the "Download CSV" button becoming enabled.

Also, even when I did add `poll` to my `OPTIONAL_APPS`, I'm still not able to see an actual CSV get generated. The celery worker doesn't seem to start any new task when I click the "Export Results to CSV" button. Any idea what I'm missing?

File: tests/unit/test_xblock_poll.py (hunk 6 above)
Comment (UmanShahzad) [reply tests/unit/test_xblock_poll.py]: https://www.python.org/dev/peps/pep-0008/#imports

Comment (UmanShahzad) [reply tests/unit/test_xblock_poll.py]: Oh nvm, I see Brad already mentioned that below.

**Comment by cgopalan** [comment]
@bradenmacdonald where are you checking for the CSVs? It should be in `/tmp/edx-s3/grades/<id>/`

**Comment by bradenmacdonald** [comment]
@cgopalan The folder `/tmp/edx-s3/` doesn't even exist on my devstack - am I missing something?

**Comment by cgopalan** [comment]
@bradenmacdonald its created when you hit "Export".

**Comment by bradenmacdonald** [comment]
@cgopalan Well, I did click that...

**Comment by cgopalan** [comment]
@bradenmacdonald strange, its working for me. As soon as I click 'export', I can see a directtory and the file created like this:
```
edxapp@vagrant:~/edx-platform$ ls -alt /tmp/edx-s3/grades/5a0525aab94db835a763cf62557ec6c79c7ae04f/
total 12
drwxrwxr-x 2 edxapp edxapp 4096 Oct  2 16:06 .
drwxrwxr-x 3 edxapp edxapp 4096 Oct  2 16:06 ..
-rw-rw-r-- 1 edxapp edxapp  187 Oct  2 16:06 survey-data-export-2017-10-02-160625.csv
```

**Comment by bradenmacdonald** [comment]
Sorry for the false alarm - I just tried again and it seems to be working now :)

There is one issue though: The CSV export uses the auto-generated option IDs, which the user cannot control.

```
user_id,username,user_email,question,answer
5,staff,staff@example.com,What is your favorite color?,G
14,user1examplecom,user1@example.com,What is your favorite color?,G
```

Under the "answer" column, it should say "Green", not "G", since there is no easy way for course authors/administrators to view/edit the relationship between the answer IDs ("G") and the answer values ("Green"). So for any poll with a non-default question and options, the values in the answer column will be pretty much meaningless.

Otherwise, 👍  from me and I'll let @cgopalan give the final approval once the above is addressed, as I won't be around.
